Migrate from versioneer to setuptools_scm

maresb opened the pull request:
**The base of this PR is #1371.** The **first original commit** on this branch is **"Remove versioneer".**

In the first commit I remove versioneer and set the version to `"TODO"`. In the second commit, I add setuptools_scm.

I also attempt to correctly configure the nightly builds. I do this by running `aesara.misc.prepare_nightly_build` to modify `pyproject.toml` in the nightly build, updating the following entries:

```toml
[project]
name = "aesara-nightly"
dynamic = []
version = "[generated from aesara.misc.prepare_nightly_build:nightly_version_number]"

[tool.setuptools_scm]
(deleted)
```
This should be tested, but I'm reasonably confident that it will work properly, and I raise an error if the version number has an incorrect form.

One strange quirk which I don't quite understand is that in order to type-check `aesara.misc.prepare_nightly_build` I added the `types-toml` package to the mypy pre-commit config, and afterwards it raised an error about an unused `type: ignore` in `aesara/scan/op.py` L3430. Removing the `type: ignore` resolved the error. :man_shrugging: 

**Thank you for opening a PR!**

Here are a few important guidelines and requirements to check before your PR can be merged:
+ [X] There is an informative high-level description of the changes.
+ [X] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [X] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [X] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [X] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.
+ [ ] There are tests covering the changes introduced in the PR.

Don't worry, your PR doesn't need to be in perfect order to submit it.  As development progresses and/or reviewers request changes, you can always [rewrite the history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) of your feature/PR branches.

If your PR is an ongoing effort and you would like to involve us in the process, simply make it a [draft PR](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-requests#draft-pull-requests).

brandonwillard reviewed (commented):
In general, this looks good; just a few comments/corrections.

Aside from that, what other approaches are there for the nightly build process, because generating a custom `pyproject.toml` seems like it could increase our maintenance overhead.

maresb commented:
@brandonwillard, how would you feel about first merging #1371? Otherwise this PR has lots of big commits.

Thanks for the review! Looks great at first glance. Will go through it in detail now...

maresb commented on aesara/scan/op.py:
```diff
@@ -3427,7 +3427,7 @@ def profile_printer(
 
 
-@op_debug_information.register(Scan)  # type: ignore[has-type]
+@op_debug_information.register(Scan)
```
Oh yikes, that looks tricky! Thanks for the context. I will need to follow up here to try and understand better what's going on.

brandonwillard replied:
Yeah, I would _really_ like to fix that problem at the source, but it was/is annoyingly elusive.

maresb replied:
Working around this with `warn_unused_ignores = false` as below, but I'll still take a stab at understanding what's really going on.

maresb replied:
Unfortunately I haven't gotten a chance to look into this yet. The issue no longer creeps up here since readding `warn_unused_ignores = false` in #1376. Is this blocking the merge?

maresb commented on aesara/version.py:
```diff
@@ -1,27 +1,31 @@
+def __getattr__(name):
+    # (Called when the module attribute is not found.)
+    if name in deprecated_names:
+        raise RuntimeError(
```
It cannot be a warning unless I reimplement the aforementioned variables so that execution can proceed. If you think that those variables are part of Aesara's public interface and shouldn't be removed, then I'll try to reimplement them, but I wanted to default to skipping work that might be unnecessary.

brandonwillard replied:
Either one is fine; I just wasn't sure which it was supposed to be, given that both type/class objects were present.

maresb replied:
Great! Sorry for the confusion; having `DeprecationWarning` there was purely an error on my part.

brandonwillard commented:
> @brandonwillard, how would you feel about first merging #1371? Otherwise this PR has lots of big commits.

Yeah, I'll look at that next.

maresb commented on readthedocs.yml:
```diff
@@ -7,3 +7,10 @@ build:
   os: "ubuntu-20.04"
   tools:
     python: "mambaforge-4.10"
+  jobs:
+    post_checkout:
+      # This is necessary for setuptools_scm to properly read the tags. The default
+      # depth of 50 often leads to 'assert version is not None' AssertionError during
+      # the pip install build process. Alternatively, we could use
+      # 'git fetch --unshallow', but that is rather intensive.
+      - git fetch --depth 1000
```
My understanding is that we compute the version by looking through the git history to find the latest matching git tag. The default commands (see the RTD CI logs) download a truncated history. This command runs afterwards to download more. I chose 1000 because it's way more than I think we would ever need, and way smaller than the whole commit history. What Ida at don't understand is why it was no prob w versioneer

maresb replied:
Are you still concerned about this? I think `--depth 1000` is better than `--unshallow` (i.e. depth=\infty) due to bandwidth/CI time considerations, and 1000 is way more than we could conceivably need.

brandonwillard replied:
No, that's not a real concern.

maresb replied:
I have a vague suspicion about what might be going on here...

In case I'm working from a fork, then it's possible that my fork might not have all the tags. (Git is sometimes a bit weird about synchronizing tags, although mainly in the push direction.) Then `setuptools_scm` may not have the info to compute the version.

Typing this out, it doesn't make much sense, but at some point during testing I somehow ended up installing a very "old" version of aesara simply because the tags were old. I don't have any actionable ideas, this is just something to think about.

maresb commented on pyproject.toml:
```diff
@@ -0,0 +1,391 @@
+files = ["aesara", "tests"]
+plugins = ["numpy.typing.mypy_plugin"]
+
+[[tool.mypy.overrides]]
+module = ["doc.*"]
+ignore_errors = true
+check_untyped_defs = false
```
TODO: Why didn't the `setup.cfg` → `pyproject.toml` tool convert the `warn_unused_ignores = false` lines?

maresb replied:
(Actually this should be done in #1371.)

maresb commented on pyproject.toml:
```diff
@@ -0,0 +1,391 @@
+    "aesara/link/jax/jax_dispatch.py",
+    "aesara/graph/toolbox.py",
+    "aesara/scalar/basic_scipy.py",
+]
```
Added

maresb commented:
TODO: see if any of the pre-defined setuptools-scm formats are pushable to PyPI. If so, eliminate the nightly script.

maresb commented:
Details about the construction of version numbers are as follows. But this may be superfluous given the suggestion proposed in the next comment...

There are two parts to a setuptools_scm version number: the "version scheme" followed by the "local scheme".

These are both defined in terms of entry points, so the options are obtainable with

```python
from setuptools_scm import get_version
from importlib.metadata import entry_points

version_schemes = [
    ep.name for ep in entry_points(group="setuptools_scm.version_scheme")
]
local_schemes = [
    ep.name for ep in entry_points(group="setuptools_scm.local_scheme")
]
print(f"{version_schemes=}\n\nlocal_schemes={local_schemes}")
```

```python
version_schemes=['calver-by-date', 'guess-next-dev', 'no-guess-dev', 'post-release', 'python-simplified-semver', 'release-branch-semver']

local_schemes=['dirty-tag', 'no-local-version', 'node-and-date', 'node-and-timestamp']
```

To first inspect just the `version_scheme`, we can set `local_scheme="no-local-version"`:

```python
print(
    {
        version_scheme: get_version(
            version_scheme=version_scheme, local_scheme="no-local-version"
        )
        for version_scheme in version_schemes
    }
)
```

```python
{'calver-by-date': '22.12.25.0.dev97', 'guess-next-dev': '2.8.10.dev97', 'no-guess-dev': '2.8.9.post1.dev97', 'post-release': '2.8.9.post97', 'python-simplified-semver': '2.8.10.dev97', 'release-branch-semver': '2.9.0.dev97'}
```

Note how all of the possibilities except for `calver-by-date`, `no-guess-dev`, and `post-release` increment the version number in some way.

The most concise possibility is `post-release`, so next we fix this and look at `local_scheme`, which depends on whether or not the repository is clean:

```python
print("Clean:")
print(
    {
        local_scheme: get_version(
            version_scheme="post-release", local_scheme=local_scheme
        )
        for local_scheme in local_schemes
    }
)
!echo >> pyproject.toml
print("Dirty:")
print(
    {
        local_scheme: get_version(
            version_scheme="post-release", local_scheme=local_scheme
        )
        for local_scheme in local_schemes
    }
)
!git checkout -q pyproject.toml
```

```python
Clean:
{'dirty-tag': '2.8.9.post97', 'no-local-version': '2.8.9.post97', 'node-and-date': '2.8.9.post97+g1945ad8cb', 'node-and-timestamp': '2.8.9.post97+g1945ad8cb'}
Dirty:
{'dirty-tag': '2.8.9.post97+dirty', 'no-local-version': '2.8.9.post97', 'node-and-date': '2.8.9.post97+g1945ad8cb.d20221226', 'node-and-timestamp': '2.8.9.post97+g1945ad8cb.d20221226110450'}
```

maresb commented:
Why not just do `version_scheme="post-release", local_scheme="no-local-version"`. Then the version number will look like `2.8.9.post97`, where `97` is the number of commits on main since the last release. (`git checkout @~97` switches to `rel-2.8.9`.) Rather than running the job nightly, we can run it on merge-to-main, which is probably more useful anyways.

This could be accomplished in a GH Actions workflow with a few simple `dasel` commands to set `version_scheme` and `local_scheme` in `pyproject.toml`, and would eliminate the need for Python scripts.

Then `aesara-nightly` is no longer a fitting name, so either we leave it as-is for "historical reasons" or we come up with something else.

maresb commented:
Just pushed a quick implementation of the idea of my last post (only tested locally).

maresb commented:
Rebased on #1376

maresb commented:
Rebased on main

brandonwillard commented on .github/workflows/dev-builds.yml:
```diff
@@ -0,0 +1,37 @@
+      - name: Install dependencies and customize pyproject.toml
+        run: |
+          # Download dasel to modify pyproject.toml
+          curl -sSLf https://github.com/TomWright/dasel/releases/download/v2.0.2/dasel_linux_amd64 \
+            -L -o /tmp/dasel && chmod +x /tmp/dasel
+
+          # Modify pyproject.toml to set the nightly version in the form of
+          # x.y.z.postN, where N is the number of commits since the last release
+          /tmp/dasel put -f pyproject.toml project.name -v aesara-nightly
+          /tmp/dasel put -f pyproject.toml tool.setuptools_scm.version_scheme -v post-release
+          /tmp/dasel put -f pyproject.toml tool.setuptools_scm.local_scheme -v no-local-version
```
This looks neat!

maresb replied:
Great, then let's keep this in favor of the Python script-based approach. (Will squash the commit which introduces this.)

maresb replied:
Squashed. What do you think about the dev builds being called `aesara-nightly` while being built on merge-to-main?

brandonwillard replied:
That sounds good.

brandonwillard reviewed (approved):
Looks good.  As always, it's hard to tell when/if these kinds of changes will work right away, but we can always follow up!

maresb commented:
This requires at least one more iteration, since `setuptools_scm` completely replaces the selection of files in the package (Git-driven). See ea65078 for all the changes. In particular,

* I noticed that `aesara/scan/scan_perform.pyx` is missing from the current distributions. It looks important, so I added it.

* I noticed that two files were being included under `tests/link/c/c_code` by virtue of having the `.c` and `.h` extensions. This looks wrong, so I removed it.

Additionally, I noticed that `aesara-cache` is being installed in a top-level module called `bin`. This seems very undesirable, as it should probably instead go under `aesara.bin`. I will make a subsequent PR to address this.

maresb commented:
I had a look and the test is failing due to more setuptools nastiness... without the `[tool.setuptools.package-data]` block which I removed from `pyproject.toml`, `setuptools_scm` manages to create a sdist containing `.c` files. This much I checked. But the big surprise comes during installation, and without that block those files are not installed. (If I edit the `pyproject.toml` within the `.tar.gz` and readd the block, then it installs correctly.) Looks like I just have to restore the block (and add `*.pyx` to get `scan_perform.pyx`)

maresb commented:
Just rebased on main. All checks were green.

rlouf commented:
One question: won't this cause issues when running the tests locally if the user hasn't built the package locally since there won't be no `_version.py` file?

maresb commented:
The `_version.py` file is generated not only by building but also by installing. So as long as the user runs `pip install [--editable] .` as they should, it should work.

What can however go wrong is that for an `--editable` install, if the version number changes, then you have to rerun `pip install -e .` in order to trigger an update to `_version.py`. Also if you have a clone which is somehow missing some release tags, then the version may also be incorrect. (That happened to me at some point as I described in https://github.com/aesara-devs/aesara/pull/1375#discussion_r1060207816, but unfortunately I don't remember exactly how.)

maresb commented:
I'm not sure what to think of this:

```
Warning: Performance alert! Previous value was 61258.103085955845 and current value is 28070.87678331834. It is 2.182265397650837x worse than previous exceeding a ratio threshold 2
Warning: Performance alert! Previous value was 1.3921410587885548 and current value is 0.6737363168561912. It is 2.066299565510444x worse than previous exceeding a ratio threshold 2
Error: # :warning: **Performance Alert** :warning:
```

maresb commented:
Hmm, dependabot did bump the version on the performance tests. I wonder if that has anything to do with it.

brandonwillard commented:
> Hmm, dependabot did bump the version on the performance tests. I wonder if that has anything to do with it.
> 
> On the other hand, the benchmarks succeeded in #1384. Maybe this simply deployed to slow runners.

Looks like it's one of the JAX benchmarks, so it could be due to a number of things (e.g. compilation, test seeding, etc.)  I'll put in a PR to at least guarantee that the test data is the same every time.

maresb commented:
The benchmarks failed again.

I wonder if the automerge failed to properly await the coverage results and prematurely merged https://github.com/aesara-devs/aesara/pull/1382#issuecomment-1375269170.

Since I believe I know what's wrong, I prepended a commit to this PR to hopefully fix it. @brandonwillard, do you still approve of a rebase (no squash) merge with the extra commit?

maresb commented:
Is it possible to rerun the "Benchmarks" test without rerunning the whole thing?

maresb commented:
@brandonwillard, tests were all green before. Are we okay to automerge (rebase) now?

maresb commented:
@brandonwillard thanks, now I'm just missing an approving review. (It wouldn't allow me to self-approve this time.)

maresb commented:
# Addendum

This PR unintentionally dropped support for `git archive`. See https://github.com/aesara-devs/aesara/pull/1405 for a fix.
